src/ResourceManagement/Network/AzureFirewallImpl.cs

```diff
@@ -0,0 +1,1298 @@
+        }
+
+        private AzureFirewallImpl WithDestinationFullQualifiedDomainNamesInNetworkRule(IList<string> domainNames)
+        {
+            networkRule.DestinationFqdns = domainNames;
+            return this;
+        }
+
+        private AzureFirewallImpl WithDestinationIpGroupInNetworkRule(string destinationIpGroup)
+        {
+            if (networkRule.DestinationIpGroups == null)
+            {
+                networkRule.DestinationIpGroups = new List<string>();
+            }
+            networkRule.DestinationIpGroups.Add(destinationIpGroup);
+            return this;
+        }
+
+        private AzureFirewallImpl WithDestinationIpGroupsInNetworkRule(IList<string> destinationIpGroups)
+        {
+            networkRule.DestinationIpGroups = destinationIpGroups;
+            return this;
+        }
+
+        private AzureFirewallImpl WithDestinationPortInNatRule(string destinationPort)
+        {
+            if (natRule.DestinationPorts == null)
+            {
+                natRule.DestinationPorts = new List<string>();
+            }
+            natRule.DestinationPorts.Add(destinationPort);
+            return this;
+        }
+
+        private AzureFirewallImpl WithDestinationPortsInNatRule(IList<string> destinationPorts)
+        {
+            natRule.DestinationPorts = destinationPorts;
+            return this;
+        }
+
+        private AzureFirewallImpl WithDestinationPortInNetworkRule(string destinationPort)
+        {
+            if (networkRule.DestinationPorts == null)
+            {
+                networkRule.DestinationPorts = new List<string>();
+            }
+            networkRule.DestinationPorts.Add(destinationPort);
+            return this;
+        }
+
+        private AzureFirewallImpl WithDestinationPortsInNetworkRule(IList<string> destinationPorts)
+        {
+            networkRule.DestinationPorts = destinationPorts;
+            return this;
+        }
+
+        private AzureFirewallImpl WithDnatActionType()
+        {
+            natRuleCollection.Action = new AzureFirewallNatRCAction(AzureFirewallNatRCActionType.Dnat);
+            return this;
+        }
+
+        private AzureFirewallImpl WithFirewallPolicy(string firewallPolicyId)
+        {
+            Inner.FirewallPolicy = new SubResource(firewallPolicyId);
+            return this;
+        }
+
+        private AzureFirewallImpl WithFullQualifiedDomainNameTagInApplicationRule(string tag)
+        {
+            if (applicationRule.FqdnTags == null)
+            {
+                applicationRule.FqdnTags = new List<string>();
+            }
+            applicationRule.FqdnTags.Add(tag);
+            return this;
+        }
+
+        private AzureFirewallImpl WithFullQualifiedDomainNameTagsInApplicationRule(IList<string> tags)
+        {
+            applicationRule.FqdnTags = tags;
+            return this;
+        }
+
+        private AzureFirewallImpl WithHttpProtocolInApplicationRule(int? port)
```

Comment:
What happens when port is null?

Reply:
As per the [Azure Firewall Application Rule Protocol](https://docs.microsoft.com/en-us/rest/api/firewall/azurefirewalls/createorupdate#azurefirewallapplicationruleprotocol), the port is optional parameter. In the interface, I placed port as optional input arg.